api/src/main/java/org/apache/iceberg/RowKey.java

```diff
@@ -0,0 +1,179 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Set;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Row key of a table.
+ * <p>
+ * Row key is a definition of table row uniqueness,
+ * similar to the concept of primary key in a relational database system.
+ * A row should be unique in a table based on the values of each {@link RowKeyIdentifierField}.
+ * Iceberg itself does not enforce row uniqueness based on this key.
+ * It is leveraged by operations such as streaming upsert.
+ */
+public class RowKey implements Serializable {
+
+  private static final RowKey NOT_IDENTIFIED = new RowKey(new Schema(), Sets.newHashSet());
+
+  private final Schema schema;
+  private final RowKeyIdentifierField[] identifierFields;
+
+  private transient volatile Set<RowKeyIdentifierField> identifierFieldSet;
+
+  private RowKey(Schema schema, Set<RowKeyIdentifierField> identifierFields) {
+    this.schema = schema;
+    this.identifierFields = identifierFields.toArray(new RowKeyIdentifierField[0]);
+  }
+
+  /**
+   * Returns the {@link Schema} referenced by the row key
+   */
+  public Schema schema() {
+    return schema;
+  }
+
+  /**
+   * Return the set of {@link RowKeyIdentifierField} in the row key
+   * <p>
+   * @return the set of fields in the row key
+   */
+  public Set<RowKeyIdentifierField> identifierFields() {
+    return lazyIdentifierFieldSet();
+  }
+
+  private Set<RowKeyIdentifierField> lazyIdentifierFieldSet() {
+    if (identifierFieldSet == null) {
+      synchronized (this) {
+        if (identifierFieldSet == null) {
+          identifierFieldSet = ImmutableSet.copyOf(identifierFields);
+        }
+      }
+    }
+
+    return identifierFieldSet;
+  }
+
+  /**
+   * Returns the default row key that has no field
+   */
+  public static RowKey notIdentified() {
+    return NOT_IDENTIFIED;
+  }
+
+  /**
+   * Returns true if the row key is the default one with no field
+   */
+  public boolean isNotIdentified() {
+    return identifierFields.length < 1;
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    } else if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+
+    RowKey that = (RowKey) other;
+    return identifierFields().equals(that.identifierFields());
+  }
+
+  @Override
+  public int hashCode() {
+    return Arrays.hashCode(identifierFields);
```

Comment:
Is this correct ?  Here different fields order will produce different hash code ?